mock-hpos-api/index.js

```diff
@@ -0,0 +1,147 @@
+const express = require('express')
+const cors = require('cors')
+const bodyParser = require('body-parser')
+const stringify = require('fast-json-stable-stringify')
+const _ = require('lodash')
+const HpAdminKeypair = require('@holo-host/hp-admin-keypair').HpAdminKeypair
+const { signPayload, hashString } = require('./authUtils')
+const defaultResponse = require('./defaultResponse')
+
+const HC_PUBKEY = '5m5srup6m3b2iilrsqmxu6ydp8p8cr0rdbh4wamupk3s4sxqr5'
+
+const API_ROOT = '/api/v1'
+
+function removeApiRoot (path) {
+  if (path.indexOf(API_ROOT) == 0) {
+    return path.slice(API_ROOT.length);
+  } else {
+    return path
+  }
+}
+
+function generateResponseKey (method, path, data) {
+  return `${method}:${removeApiRoot(path)}:${JSON.stringify(data)}`
+}
+
+// NEXT_PATH is used to key a next response in place of an actual path. It's required to be different from any value that could come from req.path
+const NEXT_PATH = ' . next . '
+
+const NEXT_RESPONSE_KEY = generateResponseKey('', NEXT_PATH, {})
+
+class MockHposApi {
+  constructor(port, authEmail, authPassword) {
+    this.port = port
+    this.shouldCheckAuth = !!authEmail && !!authPassword
+    this.authEmail = authEmail
+    this.authPassword = authPassword
+    
+    this.initializeResponses()
+  }
+
+  static async start (port, authEmail, authPassword) {
+    const mockHposApi = new MockHposApi(port, authEmail, authPassword)
+
+    const app = express()
+
+    app.use(cors())
+    app.use(bodyParser.json())
+    app.use(mockHposApi.checkAuth.bind(mockHposApi))
+    app.use(mockHposApi.handleRequest.bind(mockHposApi))
+
+    await mockHposApi.startServer(app)
+
+    return mockHposApi
+  }
+
+  // this is more convoluted than I'd like because I want init to be a blocking call, so have to uncallback app.listen
+  startServer (app) {
+    return new Promise((resolve, reject) => {
+      this.server = app.listen(this.port, () => {
+        console.log(`Mock hpos-admin-api listening at http://localhost:${this.port}`)
+        resolve()
+      }).on('error', reject)
+    })
+  }
+
+  async checkAuth (req, res, next) {
+    if (!this.shouldCheckAuth) {
+      next()
+    } else {
+
+      const { method, path, body } = req
+
+      const keypairInstance = new HpAdminKeypair(HC_PUBKEY, this.authEmail, this.authPassword)
+
+      const signature = await signPayload(keypairInstance, method, path, body)
+
+      if (signature === req.header('x-hpos-admin-signature')) {
+        next()
```

Comment:
I don't understand this. I'm assuming that `next()` means that the next request that  will be processed as sort of "mock preparation" that sets the output of the next request to come. But it also looks like `checkAuth` is called on every incoming request. This makes me think that every request is resulting in either 401 status or just configuring the mock. What am I missing?

Reply:
`next()` here is part of the express middleware api, distinct from the 'next' method of the `MockHposApi` class. It means call the next middleware. The sequence of middleware is based on the sequence they are added using `app.use` starting at line 46. So for each request, (we run a couple of utility middleware) then we run `checkAuth` and then if the signature matches, we call `next()`, which will run the `handleRequest` middleware, and if the signature doesn't match, we shortcut the sequence and immediately return a 401.